### PR TITLE
Add snaps flow

### PR DESCRIPTION
**Pull request by daria-github**
Added snap logic to inbox — if snap is not installed but user can install it and is using Metamask, prompt them through the flow. If they reject it, redirect to old flow. If they accept it, install the snap to simplify future login flows.

### CODE REVIEW

File: src/hooks/useInitXmtpClient.ts [hunk 1]
```diff
@@ -167,23 +180,61 @@ const useInitXmtpClient = () => {
...
-          // persist client keys
-          storeKeys(address, keys);
+
+          if (window.ethereum?.isMetaMask) {
+            // Check snaps flow
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
+            const browserSupportSnaps = await Client.isSnapsReady();
+            if (browserSupportSnaps) {
+              try {
+                const result = await (
```
Comment (neekolas) [inline src/hooks/useInitXmtpClient.ts:190]: I'd like to figure out how we can move some of this logic into the SDK. We already do a fair bit to make Snaps transparent, and just move on to other Keystore Providers if Snaps are not available.

If we made the flow catch rejections of the Snap installation and just move on to the next Keystore provider, would we be able to make all this code disappear?

Comment (neekolas) [reply src/hooks/useInitXmtpClient.ts:190]: My worry is that developers are going to look at this as the canonical example of how to use Snaps in their app, and Snaps should really be transparent if we want to get adoption.

File: src/hooks/useInitXmtpClient.ts [hunk 2]
```diff
@@ -19,6 +20,18 @@ type ClientStatus = "new" | "created" | "enabled";
 
 type ResolveReject<T = void> = (value: T | PromiseLike<T>) => void;
 
+interface Ethereum {
+  request(args: {
+    method: "wallet_requestSnaps";
```
Comment (neekolas) [inline src/hooks/useInitXmtpClient.ts:25]: Think we can just make the method type `string`, since window.ethereum can be used for all kinds of other calls

**Review by neekolas — APPROVED** [review]
I've tested this in the following cases and everything works as expected. Great stuff @daria-github!

- Browser with a brand new install of Metamask
- Browser with MM that already has the Snap installed
- Arc with Coinbase wallet extension
- Coinbase wallet built in mobile browser

File: src/hooks/useInitXmtpClient.ts (hunk 2 above)
Comment (daria-github) [reply src/hooks/useInitXmtpClient.ts]: yeah that's a great point -- just updated

File: src/hooks/useInitXmtpClient.ts (hunk 1 above)
Comment (daria-github) [reply src/hooks/useInitXmtpClient.ts:190]: Hmmm — agree this probably makes more sense to live in the SDK long-term. WDYT of having it land here first just to iron out any issues that come up with it, then incorporating into the SDK once enough folks have had a chance to try it out? I could even add a comment or something to clarify that this wouldn't be the end-goal of where this should live.

Comment (neekolas) [reply src/hooks/useInitXmtpClient.ts:190]: Totally fine with that approach

**Pull request merged**